globus_action_provider_tools/data_types.py

```diff
@@ -84,31 +86,93 @@ class ActionProviderDescription:
+    )
+    label: Optional[str] = Field(
+        None,
+        description="A short human presentable description of the Action requested",
+        min_length=1,
+        max_length=64,
+    )
+    deadline: Optional[datetime.datetime] = Field(
+        None,
+        description="A timestamp indicating by which time the action must complete. The request may be rejected if the Action Provider does not expect to be able to complete the action before the deadline or if it represents a time greater than the maximum_deadline specified in the Provider Description.",
+    )
+    release_after: Optional[str] = Field(
+        None,
+        description="An ISO8601 time duration value indicating how long retention of the status of the action be retained after it reaches a completed state. Action Providers may limit the maximum value. It is recommended that Providers provide a default release_after value of approximately 30 days.",
+        regex="^P(?!$)(\d+(?:\.\d+)?Y)?(\d+(?:\.\d+)?M)?(\d+(?:\.\d+)?W)?(\d+(?:\.\d+)?D)?(T(?=\d)(\d+(?:\.\d+)?H)?(\d+(?:\.\d+)?M)?(\d+(?:\.\d+)?S)?)?$",
+    )
+    monitor_by: Set[str] = Field(
+        default_factory=list,
+        description="A list of principal URNs containing identities which are allowed to monitor the progress of the action using the status and log operations. When not provided, defaults to the user that initiated the action.",
+        regex="^urn:globus:(auth:identity|groups:id):([a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12})$",
+    )
+    manage_by: Set[str] = Field(
+        default_factory=list,
+        description="A list of principal URNs containing identities which are allowed to manage the progress of the action using the cancel and release operations. When not provided, defaults to the user that initiated the action.",
+        regex="^urn:globus:(auth:identity|groups:id):([a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12})$",
+    )
+    allowed_clients: List[str] = Field(
+        default_factory=list, regex="^(public|globus|creator|.$)$"
+    )
+
+
+class ActionStatus(BaseModel):
+    status: ActionStatusValue = Field(
+        ..., description="The current state of the Action"
+    )
+    creator_id: str = Field(
+        ...,
+        description="A URN representation of an Identity in Globus either of a user from Globus Auth or a group from Globus Groups.",
+        regex="^urn:globus:(auth:identity|groups:id):([a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12})$",
+    )
+    action_id: str = Field(
+        default_factory=shortish_id, description="The id of the Action itself"
+    )
+    start_time: str = Field(default_factory=now_isoformat)
+    label: Optional[str] = Field(
+        None,
+        description="A short human presentable description of the Action requested",
+        min_length=1,
+        max_length=64,
+    )
+    monitor_by: Set[str] = Field(
+        default_factory=list,
```

Comment:
If this is type `Set` should the `default_factory` also be `set`?

Reply:
I'm on the fence about this one, pydantic interprets the set type annotation to mean it should be a sequence of unique items, and the validation will make sure of that. The default factory is ends up being the code representation and since theyre being used in different ways, I'm not sure about the best way to go forward. Let me see if setting this to be a set factory breaks anything.